src/main/java/com/aws/greengrass/clientdevices/auth/ClientDevicesAuthService.java

```diff
@@ -200,19 +226,67 @@ protected void install() throws InterruptedException {
 
             if (whatHappened == WhatHappened.initialized || node == null) {
                 updateDeviceGroups(whatHappened, deviceGroupTopics);
-                updateCAType(caTypeTopic);
+                updateCertificateAuthority();
             } else if (node.childOf(DEVICE_GROUPS_TOPICS)) {
                 updateDeviceGroups(whatHappened, deviceGroupTopics);
-            } else if (node.childOf(CA_TYPE_TOPIC)) {
-                if (caTypeTopic.getOnce() == null) {
+            } else if (node.childOf(CERTIFICATE_AUTHORITY_TOPIC)) {
+                if (certificateAuthorityTopic.getOnce() == null) {
                     return;
                 }
-                updateCAType(caTypeTopic);
+                updateCertificateAuthority();
             }
         });
 
     }
 
+    private void updateCertificateAuthority() {
+        String caKeyPath = Coerce.toString(config.findOrDefault("",
+                CERTIFICATE_AUTHORITY_TOPIC, CA_PRIVATE_KEY_URI));
+        String caCertPath = Coerce.toString(config.findOrDefault("",
+                CERTIFICATE_AUTHORITY_TOPIC, CA_CERTIFICATE_URI));
+        Topic caType = config.lookup(CERTIFICATE_AUTHORITY_TOPIC, CA_TYPE_TOPIC);
+
+        // If the key and cert path of CA are empty, generate a CA
+        if (Utils.isEmpty(caKeyPath) || Utils.isEmpty(caCertPath)) {
+            updateCAType(caType);
+        } else {
+            // Get CA from URI and upload
+            try {
+                URI privateKeyUri = new URI(caKeyPath);
+                URI certificateUri = new URI(caCertPath);
+                KeyPair keyPair = RetryUtils.runWithRetry(GET_KEY_PAIR_RETRY_CONFIG,
+                        () -> securityService.getKeyPair(privateKeyUri, certificateUri),
+                        "get-key-pair", logger);
+                List<X509Certificate> certificateChain = RetryUtils.runWithRetry(GET_CERTIFICATE_CHAIN_RETRY_CONFIG,
+                        () -> securityService.getCertificateChain(privateKeyUri, certificateUri),
+                        "get-certificate-chain", logger);
+                certificateManager.setCA(keyPair.getPrivate(), (Certificate[]) certificateChain.toArray());
+            } catch (URISyntaxException e) {
+                logger.atError().kv(CA_PRIVATE_KEY_URI, "")
+                        .kv(CA_CERTIFICATE_URI, "")
+                        .setCause(e)
+                        .log("Exception occurred while parsing the certificate authority configuration");
+                serviceErrored(e);
+            } catch (InterruptedException | KeyLoadingException | IOException | CertificateException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)